### PR TITLE
DataSourceAzure: Use ValueError when JSONDecodeError is not available

### PR DESCRIPTION
**Pull request by anhvoms**
JSONDecodeError is only available in Python 3.5+. When it isn't available (i.e. on Python 3.4, which cloud-init still supports) use the more generic ValueError.

### CODE REVIEW

**Comment by igalic** [comment]
> Several distros that are still on python2

i'm pretty, and sure, cloud-init has gone python 3 only quite some time ago, so

> will throw tracebacks and fail provisioning if cloud-init cannot decode the response from IMDS due to JSONDecodeError is not available in python2.

so I'm surprised it's getting as far as this, to begin with

File: cloudinit/sources/DataSourceAzure.py [hunk 1]
```diff
@@ -1436,9 +1436,10 @@ def _get_metadata_from_imds(retries):
         report_diagnostic_event(msg)
         LOG.debug(msg)
         return {}
+    if
```
Comment (igalic) [inline cloudinit/sources/DataSourceAzure.py:1439]: this line is breaking all tests.

**Comment by anhvoms** [comment]
> cloud-init dropped support for python 2.
> https://lists.launchpad.net/cloud-init/msg00227.html
> https://pythonclock.org/

Yes we're aware that it did. However, there were discussions in the cloud-summit that there are distros that will continue to support (only) python2 and will cherry pick commits that suit their needs. This is meant for those distros. I don't see cloud-init dropping support for python2 being a good reason to block this PR.

**Comment by anhvoms** [comment]
> > Several distros that are still on python2
> 
> i'm pretty, and sure, cloud-init has gone python 3 only quite some time ago, so
> 
> > will throw tracebacks and fail provisioning if cloud-init cannot decode the response from IMDS due to JSONDecodeError is not available in python2.
> 
> so I'm surprised it's getting as far as this, to begin with

Cloud-init only stopped supporting python2 since 19.4, which is about 6-8 months ago from the last SRU.

**Comment by smoser** [comment]
> Yes we're aware that it did. However, there were discussions in the cloud-summit that there are distros that will continue to support (only) python2 and will cherry pick commits that suit their needs. This is meant for those distros. I don't see cloud-init dropping support for python2 being a good reason to block this PR.

Catching a more generic `ValueError` when we only intend to catch `json.decoder.JSONDecodeError` is generally not "best practice".  If the only reason we're doing it is "to make python 2 work", then we did not "drop support for python 2".

With regard to distros maintaining python 2 support, I believe that people would be open to setting up a `python2` branch, possibly with different commit requirements/ownership.  The entire intent of "dropping python 2" was to not have to think about this sort of thin, so this really does not fit in master.

**Comment by anhvoms** [comment]
> > Yes we're aware that it did. However, there were discussions in the cloud-summit that there are distros that will continue to support (only) python2 and will cherry pick commits that suit their needs. This is meant for those distros. I don't see cloud-init dropping support for python2 being a good reason to block this PR.
> 
> Catching a more generic `ValueError` when we only intend to catch `json.decoder.JSONDecodeError` is generally not "best practice". If the only reason we're doing it is "to make python 2 work", then we did not "drop support for python 2".
> 
> With regard to distros maintaining python 2 support, I believe that people would be open to setting up a `python2` branch, possibly with different commit requirements/ownership. The entire intent of "dropping python 2" was to not have to think about this sort of thin, so this really does not fit in master.

For Azure it's actually ok if we catch a more generic ValueError here and print out a warning. If you would prefer to have that I'm happy to resend another PR with a better commit message.

**Comment by OddBloke** [comment]
We have dropped support for Python 2 in cloud-init master, and as this change does make the codebase worse (albeit only very slightly, by making the exception handler less specific), it would not appropriate for inclusion in master. However, it appears that [`JSONDecodeError` was only introduced in Python 3.5](https://docs.python.org/3/library/json.html#json.JSONDecodeError) and cloud-init _does_ still support Python 3.4. If you can update the PR description to reflect that being the reason it should be included, then I think we can probably review/accept this.

(As an aside, we specifically created the [`stable-19.4`](https://github.com/canonical/cloud-init/tree/stable-19.4) branch for community members to continue to collaborate on supporting Python 2; if this were only a Python 2 compatibility issue, it would seems like a perfect candidate for inclusion there.)

**Comment by anhvoms** [comment]
> We have dropped support for Python 2 in cloud-init master, and as this change does make the codebase worse (albeit only very slightly, by making the exception handler less specific), it would not appropriate for inclusion in master. However, it appears that [`JSONDecodeError` was only introduced in Python 3.5](https://docs.python.org/3/library/json.html#json.JSONDecodeError) and cloud-init _does_ still support Python 3.4. If you can update the PR description to reflect that being the reason it should be included, then I think we can probably review/accept this.
> 
> (As an aside, we specifically created the [`stable-19.4`](https://github.com/canonical/cloud-init/tree/stable-19.4) branch for community members to continue to collaborate on supporting Python 2; if this were only a Python 2 compatibility issue, it would seems like a perfect candidate for inclusion there.)

Thanks @OddBloke , I have updated the PR description

**Comment by raharper** [comment]
I believe we have several places[1][2][3] where we attempt to import a particular exception and fallback on the more general error if it's not present:

```
try:
   import json
   json_decode_error = json.decoder.JSONDecodeError
except ImportError, AttributeError:
   json_decode_error = ValueError

```

1. cloudinit/handlers/jinja_template.py" line 7
2. cloudinit/serial.py
3. cloudinit/sources/helpers/vmware/imc/config_file.py

**Comment by anhvoms** [comment]
> try:
>    import json
>    json_decode_error = json.decoder.JSONDecodeError
> except ImportError, AttributeError:
>    json_decode_error = ValueError

Thanks @raharper. I have followed this suggestion.

File: cloudinit/sources/DataSourceAzure.py [hunk 2]
```diff
@@ -1436,9 +1435,15 @@ def _get_metadata_from_imds(retries):
         report_diagnostic_event(msg)
         LOG.debug(msg)
         return {}
+    try:
+        import json
+        json_decode_error = json.decoder.JSONDecodeError
+    except ImportError, AttributeError:
```
Comment (OddBloke) [inline cloudinit/sources/DataSourceAzure.py:1441]: I think we can simplify this a little:

```suggestion
    try:
        from json.decoder import JSONDecodeError
    except ImportError:
```

**Pull request merged**